src/compiler/crystal/macros/methods.cr

```diff
@@ -798,6 +798,18 @@ module Crystal
         interpret_argless_method(method, args) { StringLiteral.new(@value.underscore) }
       when "upcase"
         interpret_argless_method(method, args) { StringLiteral.new(@value.upcase) }
+      when "parse_type_name"
+        interpret_argless_method(method, args) do
+          raise "StringLiteral#parse_type_name cannot be called on an empty string" if @value.blank?
+
+          parser = Crystal::Parser.new @value
+          parser.next_token_skip_statement_end
+          type = parser.parse_bare_proc_type
+          parser.check :EOF
+          type
+        rescue ex : Crystal::SyntaxException
+          raise "Invalid type name: #{@value}"
```

Comment:
I think the value should probably be delimited to account for whitespace (think of a blank string, for example):

```suggestion
          raise "Invalid type name: #{@value.inspect}"
```

Reply:
@straight-shoota Technically empty strings are disallowed, but wont hurt anything